### ISSUE TITLE
Plans for supporting t.Parallel

### ISSUE THREAD

**Issue by tsenart** [fortytw2/leaktest#4]
Hello!

Thanks for this library, it'll definitely come in handy. Would you be able to document the reasons why t.Parallel isn't supported and what ideas you have for addressing the problem?

Thanks

**Comment by fortytw2** [comment]
Thanks!

So right now, `t.Parallel()` isn't supported because goroutines from tests that run in parallel tend to show up in the deferred check - making it look like the current test leaks goroutines when it doesn't.

A potential solution I've had in mind is to have a variant of `Check()` that operates on a `*testing.M`, which is what the `net/http` tests do. I haven't had the chance to test it out yet though. 

Another idea uses a sort-of context-aware, gross, package global stack-marker, but I don't think that'll pan out the way I'd like.

**Comment by tsenart** [comment]
Thanks! :-)

**Comment by fortytw2** [comment]
Can track support for this in https://github.com/fortytw2/leaktest/pull/16

Not entirely sure of this solution yet though

**Comment by pwaller** [comment]
You guys might be interested in the recently implemented https://github.com/golang/go/issues/22289. ("proposal: runtime: support tracking goroutine ancestor tracebacks").

Unfortunately I don't see a way of turning it on at runtime without spawning another process to set the environment variable to turn it on, hmm.

**Comment by bwplotka** [comment]
Looking forward for this to be done. Can I help anyhow? Any update? (:

**Comment by fortytw2** [comment]
I haven't worked on this in a long time @bwplotka, you're more than welcome to pick it up! 

What I did two years ago was start allowing this via just add a `TestMain(` helper called `CheckMain`, that implicitly supports `t.Parallel` by just checking the entire program. Never got really far / found the need for it in any of my projects or work though, so I stopped using it.

For the last year now, the codebase I work on at $dayjob uses ginkgo, which has a parallel test runner that segments tests into independent binaries during parallel execution, which makes leaktest work as normal (but we don't have a real use case for it).  

A solution based on what @pwaller noted last year would probably be the best way forward, feel free to give it a shot! You can message me on gophers slack (same username) if you want to chat about it

**Comment by bwplotka** [comment]
Awesome, will let know here if I will find time to work on this (: Thanks for tips! :+1:
